### PR TITLE
*: new `httputil.Client` and methods, & use them

### PR DESCRIPTION
**Pull request by knz**
As discussed offline, this was found while investigating CI failures due to an unreachable telemetry server.

Previously, various parts of CockroachDB were using the `http`
package's Get/Post/Head methods/functions directly. This was defective
for two reasons:

- it did not implement a request timeout, which could cause
  requests to hang forever if the server was inoperative.
- it did not obey context cancellation, which would force
  the CockroachDB server to hang forever while trying to
  quit if some HTTP server was blocking.

The last point in particular may cause excessive operator friction
when restarting a server, by forcing a node to be killed and
potentially disrupting operation.

This patch fixes the issue by introducing a new `httputil.Client`
class which implements a customizable timeout, see its constructor
`NewClientWithTimeout`.  The default instance in
`httputil.DefaultClient` uses a 5 second timeout.

It also introduces context-aware variants of the `http`
methods/functions `Get`, `Post` and `Head`.  Note that these variants
do not _actually_ obey context cancellation when the code is compiled
with Go 1.12 or prior. The new behavior only kicks in with Go 1.13 or
later (it requires the new feature `NewRequestWithContext`). However,
the usability problem should still be somewhat alleviated in Go 1.12
builds thanks to the request timeout implemented by the new
`httputil.Client`.

The following components in CockroachDB are then adapted to use the
new interface:

- avro registration for changefeed encoders: timeout, no cancellation
  (missing context.Context plumbing - left TODO)
- telemetry updates: timeout + cancellation
- binfetcher: timeout + cancellation
- cloudinfo: timeout + cancellation
- docgen: timeout, no cancellation (interactive tool)
- roachprod install: timeout, no cancellation (interactive tool)
- test server: timeout, no cancellation (tests have timeouts already)

Release note (bug fix): CockroachDB will now less likely hang in an
inconvenient/inoperative state if it attempts to access an external
HTTP server that blocks or is overloaded. A possible symptom of
the bug is a node failing to shut down upon `cockroach quit`. This
bug is present since at least version 2.0.

### CODE REVIEW

**Comment by knz** [comment]
cc @danhhz for the changefeed update
@dt for the telemetry update

**Review by tbg — APPROVED** [review]
> Reviewed 11 of 11 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42536)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Comment by knz** [comment]
Gah, `NewRequestWithContext` is a go 1.13 feature. I can't merge this until we bump the required version.

**Comment by knz** [comment]
I'll dumb it down with just the basic form.

**Comment by knz** [comment]
PTAL, especially what we expect with pre-1.13 builds (I know we're bumping the repo to 1.13 before crdb 20.1, but I'd like this patch to be back-ported to 19.2).

**Review by danhhz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42536)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @tbg)

___
*[pkg/ccl/changefeedccl/encoder.go, line 425 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42536#-Ltz7N-r7D23R5483pwi:-Ltz7N-r7D23R5483pwj:bkit1wp) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8d3b61bd4f6072787e81929b7b1e3af12b2162/pkg/ccl/changefeedccl/encoder.go#L425)):*
> ```Go
> 	}
> 
> 	// TODO(someone): connect the context to the caller to obey
> ```

I worry that this TODO will not get done. It looks like register is called by EncodeKey/EncodeValue/EncodeResolvedTimestamp. I think all callers of those three already have a context, so maybe it isn't bad to do it as part of this PR (perhaps as a separate commit).

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42536)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @tbg)

___
*[pkg/ccl/changefeedccl/encoder.go, line 425 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42536#-Ltz7N-r7D23R5483pwi:-LtzCKUP4KwzogllLukP:blir1xx) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8d3b61bd4f6072787e81929b7b1e3af12b2162/pkg/ccl/changefeedccl/encoder.go#L425)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I worry that this TODO will not get done. It looks like register is called by EncodeKey/EncodeValue/EncodeResolvedTimestamp. I think all callers of those three already have a context, so maybe it isn't bad to do it as part of this PR (perhaps as a separate commit). 
</blockquote></details>

No problem. Done.

**Review by danhhz — COMMENTED** [review]
changefeedccl lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42536#-:-LtzFVHCAduthLnpkhPb:bharjja)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @tbg)

___
*[pkg/ccl/changefeedccl/encoder.go, line 425 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/42536#-Ltz7N-r7D23R5483pwi:-LtzFRXv4XSN_GHCPnBH:bnats8d) ([raw file](https://github.com/cockroachdb/cockroach/blob/bb8d3b61bd4f6072787e81929b7b1e3af12b2162/pkg/ccl/changefeedccl/encoder.go#L425)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

No problem. Done.
</blockquote></details>

:clap:

**Comment by knz** [comment]
TFYRs!

bors r+

**Comment by knz** [comment]
The PR is now failing the linter, but there is no lint error. I think the lint test process died (maybe OOM-killed? not sure). @tbg any suggestion as to how to troubleshoot this?

**Comment by tbg** [comment]
Hmm, if something got oom killed, would we expect an error 1 here?

```
[TestLint/TestVet] TestLint/TestVet
[16:58:33]
[TestLint/TestVet] [Test Output]
=== PAUSE TestLint/TestVet
[16:58:33]
[TestLint/TestVet] TestLint/TestVet/shadow
[16:58:33]
[TestLint/TestVet] make: *** [lint] Error 1
```

Does `make lint` pass locally? I agree that this is weird. Does the problem disappear if you revert your changes to the linters?

**Comment by knz** [comment]
I found it. Somehow the test vet output gets swallowed somewhere, but it was reporting a legitimate error (which I have now fixed).

I'm going to wait for a green build now and merge afterwards.
